src/core/emotiondisasm.cpp

```diff
@@ -5,56 +5,1058 @@
+            return disasm_daddiu(instruction);
+        case 0x1A:
+            return disasm_ldl(instruction);
+        case 0x1B:
+            return disasm_ldr(instruction);
+        case 0x1C:
+            return disasm_mmi(instruction, instr_addr);
+        case 0x1E:
+            return disasm_lq(instruction);
+        case 0x1F:
+            return disasm_sq(instruction);
+        case 0x20:
+            return disasm_lb(instruction);
+        case 0x21:
+            return disasm_lh(instruction);
+        case 0x22:
+            return disasm_lwl(instruction);
+        case 0x23:
+            return disasm_lw(instruction);
+        case 0x24:
+            return disasm_lbu(instruction);
+        case 0x25:
+            return disasm_lhu(instruction);
+        case 0x26:
+            return disasm_lwr(instruction);
+        case 0x27:
+            return disasm_lwu(instruction);
+        case 0x28:
+            return disasm_sb(instruction);
+        case 0x29:
+            return disasm_sh(instruction);
+        case 0x2A:
+            return disasm_swl(instruction);
+        case 0x2B:
+            return disasm_sw(instruction);
+        case 0x2C:
+            return disasm_sdl(instruction);
+        case 0x2D:
+            return disasm_sdr(instruction);
+        case 0x2E:
+            return disasm_swr(instruction);
+        case 0x2F:
+            return "cache";
+        case 0x31:
+            return disasm_lwc1(instruction);
+        case 0x36:
+            return "TODO: lqc2";
+            break;
+        case 0x37:
+            return disasm_ld(instruction);
+        case 0x39:
+            return disasm_swc1(instruction);
+        case 0x3E:
+            return "TODO: sqc2";
+        case 0x3F:
+            return disasm_sd(instruction);
+        default:
+            return unknown_op("", instruction >> 26, 2);
     }
-    return "[UNKNOWN]";
 }
 
-/*
- * Branch disassembly uses "instr_addr", the address of the instruction in memory, to calculate the
- * new address that can be branched to.
- */
+
+string EmotionDisasm::disasm_j(uint32_t instruction, uint32_t instr_addr)
+{
+    return disasm_jump("j", instruction, instr_addr);
+}
+
+string EmotionDisasm::disasm_jal(uint32_t instruction, uint32_t instr_addr)
+{
+    return disasm_jump("jal", instruction, instr_addr);
+}
+
+
+string EmotionDisasm::disasm_jump(const string opcode, uint32_t instruction, uint32_t instr_addr)
+{
+    stringstream output;
+    uint32_t addr = (instruction & 0x3FFFFFF) << 2;
+    addr += (instr_addr + 4) & 0xF0000000;
+    output << "$" << setfill('0') << setw(8) << hex << addr;
+
+    return opcode + " " + output.str();
+}
+
 string EmotionDisasm::disasm_bne(uint32_t instruction, uint32_t instr_addr)
 {
-    //Use stringstream only when putting integers into the disassembly.
+    return disasm_bequality("bne", instruction, instr_addr);
+}
+
+
+string EmotionDisasm::disasm_bequality(string opcode, uint32_t instruction, uint32_t instr_addr)
```

Comment:
Better name would be ```disasm_branch```

Reply:
I changed the name to `disasm_branch_equality` because there is an other disassembly method (now called) `disasm_branch_inequality` which takes care of BGTZ and BLTZ, etc.